File: day01/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func readInput(filePath string) string {
	data, err := ioutil.ReadFile(filePath)
	check(err)
	return string(data)
}

func countIncreases(measurements []int, windowSize int) int {
	count := 0
	for i := 0; i < len(measurements)-windowSize; i++ {
		if measurements[i+windowSize] > measurements[i] {
			count++
		}
	}
	return count
}

func main() {
	args := os.Args[1:]
	if len(args) != 1 {
		fmt.Println("Input path needs to be specified, no other arguments are accepted")
		os.Exit(1)
	}
	inputFile := args[0]

	data := readInput(inputFile)
	inputs := strings.Split(strings.TrimRight(data, "\n"), "\n")
	measurements := make([]int, len(inputs))
	for i, v := range inputs {
		entry, err := strconv.Atoi(v)
		check(err)
		measurements[i] = entry
	}

	fmt.Println("Part one:", countIncreases(measurements, 1))
	fmt.Println("Part two:", countIncreases(measurements, 3))
}

File: day01/main.lua
function count_increases(measurements, window_size)
  local count = 0
  for i=1,#measurements-window_size,1 do
    if measurements[i + window_size] > measurements[i] then
      count = count + 1
    end
  end
  return count
end

if #arg ~= 1 then
  print("Input path needs to be specified, no other arguments are accepted")
  os.exit(1)
end
local input_path = arg[1]

local measurements = {}
local file = io.open(input_path)
if not file then
  print("Could not find file " .. input_path)
  os.exit(1)
end
local i = 1
while true do
  input = file:read()
  if input then
    measurements[i] = tonumber(input)
    i = i + 1
  else
    break
  end
end
file:close()

print("Part one: " .. count_increases(measurements, 1))
print("Part two: " .. count_increases(measurements, 3))
